jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponse.java

```diff
@@ -249,136 +271,162 @@ protected void onCompleteFailure(Throwable x)
         @Override
         protected Action process() throws Exception
         {
-            // If we have no deflater
-            if (_deflaterEntry == null)
+            if (LOG.isDebugEnabled())
+                LOG.debug("GzipBufferCB.process(): _last={}, _buffer={}, _content={}", _last, BufferUtil.toDetailString(_buffer), BufferUtil.toDetailString(_content));
+
+            GZState gzstate = _state.get();
+
+            // Are we finished?
+            if (gzstate == GZState.FINISHED)
             {
                 // then the trailer has been generated and written below.
                 // We have finished compressing the entire content, so
                 // cleanup and succeed.
-                if (_buffer != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_buffer);
-                    _buffer = null;
-                }
-                if (_copy != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_copy);
-                    _copy = null;
-                }
+                cleanup();
                 return Action.SUCCEEDED;
             }
 
             // If we have no buffer
             if (_buffer == null)
             {
-                // allocate a buffer and add the gzip header.
                 _buffer = getRequest().getComponents().getByteBufferPool().acquire(_bufferSize, false);
-                BufferUtil.fill(_buffer, GZIP_HEADER, 0, GZIP_HEADER.length);
+                // Per RFC-1952, GZIP is LITTLE_ENDIAN
+                _buffer.order(ByteOrder.LITTLE_ENDIAN);
+                BufferUtil.flipToFill(_buffer);
```

Comment:
This is a slightly unconventional way to use BufferUtil, i.e. to flip to fill mode and leave it that way for some time.      Can you make it clear in the javadoc and/or argument names that the `compressing` and `finishing` methods expect to be called with a buffer in fill mode.  Also a comment here about who/when/where the buffer is flipped to flush would be handy.

Reply:
I should rename `_buffer` to `_outputBuffer` too.